File: myExpenses/src/main/java/org/totschnig/myexpenses/viewmodel/TransactionListViewModel.java
package org.totschnig.myexpenses.viewmodel;

import android.app.Application;
import android.content.ContentUris;
import android.net.Uri;

import com.squareup.sqlbrite3.BriteContentResolver;
import com.squareup.sqlbrite3.SqlBrite;

import org.totschnig.myexpenses.model.Account;
import org.totschnig.myexpenses.provider.TransactionProvider;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.MutableLiveData;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

public class TransactionListViewModel extends AndroidViewModel {

  private BriteContentResolver briteContentResolver;
  private Disposable disposable;

  private final MutableLiveData<Account> account = new MutableLiveData<>();

  public TransactionListViewModel(@NonNull Application application) {
    super(application);
    briteContentResolver = new SqlBrite.Builder().build().wrapContentProvider(application.getContentResolver(), Schedulers.io());
  }

  public MutableLiveData<Account> getAccount() {
    return account;
  }

  @Override
  protected void onCleared() {
    dispose();
  }

  private void dispose() {
    if (disposable != null && !disposable.isDisposed()) {
      disposable.dispose();
    }
  }

  public void loadAccount(long accountId) {
    dispose();
    final Uri base = accountId > 0 ? TransactionProvider.ACCOUNTS_URI : TransactionProvider.ACCOUNTS_AGGREGATE_URI;
    disposable = briteContentResolver.createQuery(ContentUris.withAppendedId(base, accountId),
        Account.PROJECTION_EXTENDED, null, null, null, true)
        .mapToOne(Account::fromCursor)
        .subscribe(TransactionListViewModel.this.account::postValue);
  }
}

File: myExpenses/src/main/java/org/totschnig/myexpenses/viewmodel/TransactionListViewModel.kt
package org.totschnig.myexpenses.viewmodel

import android.app.Application
import android.os.Parcelable
import androidx.lifecycle.SavedStateHandle
import app.cash.copper.flow.mapToOne
import app.cash.copper.flow.observeQuery
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.filterNotNull
import kotlinx.coroutines.withContext
import kotlinx.parcelize.Parcelize
import org.totschnig.myexpenses.db2.FLAG_NEUTRAL
import org.totschnig.myexpenses.db2.asCategoryType
import org.totschnig.myexpenses.db2.tagMapFlow
import org.totschnig.myexpenses.model.CurrencyUnit
import org.totschnig.myexpenses.model.Grouping
import org.totschnig.myexpenses.provider.DataBaseAccount.Companion.uriBuilderForTransactionList
import org.totschnig.myexpenses.provider.DatabaseConstants.KEY_CATID
import org.totschnig.myexpenses.provider.DatabaseConstants.KEY_DISPLAY_AMOUNT
import org.totschnig.myexpenses.provider.DatabaseConstants.WHERE_NOT_SPLIT
import org.totschnig.myexpenses.provider.DatabaseConstants.WHERE_NOT_VOID
import org.totschnig.myexpenses.provider.DbUtils
import org.totschnig.myexpenses.provider.effectiveTypeExpression
import org.totschnig.myexpenses.provider.effectiveTypeExpressionIncludeTransfers
import org.totschnig.myexpenses.viewmodel.data.Transaction2

const val KEY_LOADING_INFO = "loadingInfo"

class TransactionListViewModel(
    application: Application,
    private val savedStateHandle: SavedStateHandle
) : ContentResolvingAndroidViewModel(application) {

    val loadingInfo
        get() = savedStateHandle.get<LoadingInfo>(KEY_LOADING_INFO)!!

    @Parcelize
    data class LoadingInfo(
        val accountId: Long,
        val currency: CurrencyUnit,
        val label: String,
        val catId: Long = 0,
        val grouping: Grouping? = null,
        val groupingClause: String? = null,
        val groupingArgs: Array<String> = emptyArray(),
        val type: Boolean? = null,
        val aggregateNeutral: Boolean = false,
        val withTransfers: Boolean = false,
        val icon: String? = null,
        val withNewButton: Boolean = false,
        val color: Int? = null
    ) : Parcelable

    val sum: Flow<Long>
        get() = with(loadingInfo) {
            val (selection, selectionArgs) = selectionInfo
            contentResolver.observeQuery(
                transactionUri, arrayOf("sum(${KEY_DISPLAY_AMOUNT})"), selection, selectionArgs
            ).mapToOne {
                it.getLong(0)
            }
        }

    private val transactionUri
        get() = uriBuilderForTransactionList(
            loadingInfo.accountId,
            loadingInfo.currency.code,
            shortenComment = true,
            extended = false
        ).apply {
            if (loadingInfo.catId != 0L) {
                appendQueryParameter(KEY_CATID, loadingInfo.catId.toString())
            }
        }.build()


    val transactions: Flow<List<Transaction2>>
        get() = with(loadingInfo) {
            val (selection, selectionArgs) = selectionInfo
            combine(
                flow = contentResolver.tagMapFlow,
                flow2 = contentResolver.observeQuery(
                    transactionUri,
                    Transaction2.projection(
                        accountId,
                        Grouping.NONE,
                        prefHandler,
                        extended = false
                    ),
                    selection,
                    selectionArgs
                )
            ) { tags, query ->
                withContext(Dispatchers.IO) {
                    query.run()?.use { cursor ->
                        buildList {
                            while (cursor.moveToNext()) {
                                add(
                                    Transaction2.fromCursor(
                                        currencyContext,
                                        cursor,
                                        tags,
                                        currency
                                    )
                                )
                            }
                        }
                    }
                }
            }.filterNotNull()
        }

    private val selectionInfo: Pair<String, Array<String>>
        get() = with(loadingInfo) {
            val selectionParts = mutableListOf<String>()
            val selectionArgs = mutableListOf<String>()
            selectionParts += WHERE_NOT_VOID
            if (catId == 0L) {
                selectionParts += WHERE_NOT_SPLIT
            }
            groupingClause?.takeIf { it.isNotEmpty() }?.let {
                selectionParts += it
                selectionArgs.addAll(groupingArgs)
            }
            val typeWithFallback = DbUtils.typeWithFallBack(prefHandler)
            if (type != null) {
                val typeExpression = when {
                    aggregateNeutral -> "$typeWithFallback IN (${type.asCategoryType}, $FLAG_NEUTRAL)"
                    withTransfers -> effectiveTypeExpressionIncludeTransfers(typeWithFallback) + " = " + type.asCategoryType
                    else -> effectiveTypeExpression(typeWithFallback) + " = " + type.asCategoryType
                }
                selectionParts += typeExpression
            }
            selectionParts.joinToString(" AND ") to selectionArgs.toTypedArray()
        }
}

